### ISSUE TITLE
Feature Request: Schema file

### ISSUE THREAD

**Issue by factoidforrest** [knex/knex#1086]
I'm aware there is a new migration system in the works, but I was thinking a schema file could be useful, in addition to the migrations.  Rails does this to help you see how your database currently is without hunting through migrations.  It could be regenerated on every migration.

**Comment by rhys-vdw** [comment]
What is the purpose of the schema file? I assume that it is there to build your schema on a new deployment in a single operation (rather than running all the migrations).

I expect most DBMSes will have a built in function to do this. For instance PostgreSQL has `pg_dump --schema-only`. So I'm guessing that we could just call the appropriate command via command line, saving the file in some place configured in the `knexfile`. Do you know how Rails does it?

**Comment by blah238** [comment]
@tgriesser recently proposed something similar over in the Bookshelf repo here: https://github.com/tgriesser/bookshelf/issues/1019

It sounds like you are looking for something more automated, so what format were you thinking the file would take?

**Comment by danielgindi** [comment]
You can use this https://github.com/danielgindi/knex-schema-builder

**Comment by light24bulbs** [comment]
Huh, that's interesting.  Can it generate a file like this:  

``` ruby
ActiveRecord::Schema.define(:version => 11) do

  create_table "categories", :force => true do |t|
    t.string   "title"
    t.text     "description"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "categories_links", :id => false, :force => true do |t|
    t.integer "category_id"
    t.integer "link_id"
  end

  create_table "links", :force => true do |t|
    t.string   "url"
    t.string   "title"
    t.text     "description"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "messages", :force => true do |t|
    t.string   "name"
    t.string   "email"
    t.string   "company"
    t.string   "phone"
    t.string   "subject"
    t.text     "body"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "pages", :force => true do |t|
    t.string   "name"
    t.string   "title"
    t.text     "body"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.boolean  "admin"
    t.integer  "parent_id"
    t.string   "navlabel"
    t.integer  "position"
    t.boolean  "redirect"
    t.string   "action_name"
    t.string   "controller_name"
  end

  create_table "users", :force => true do |t|
    t.string   "login"
    t.string   "email"
    t.string   "crypted_password",          :limit => 40
    t.string   "salt",                      :limit => 40
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "remember_token"
    t.datetime "remember_token_expires_at"
  end

end
```

**Comment by danielgindi** [comment]
There's more to it than that, your schema file contains the indexes and
keys also, and if there are any truggers etc. to run when initializing a
database then that should be in there too. The schema is JSON.
There are upgrade schemas too, between bersions.

I'm working on creating a schema file from an existing db, which is complex
especially when there are so many db engines, but almost there with mysql.

‏
>
>

**Comment by light24bulbs** [comment]
Cool!  Yeah, it would be nice if it would auto generate.  I guess what I want is just a way to see all the DB info in one file, instead of a boatload of migrations.

**Comment by mbavio** [comment]
Any updates on this one? It could be really handy to have the schema file to be updated automatically.

**Comment by light24bulbs** [comment]
Agreed.  It's almost a form of self documenting code.  Just something for the user to read so they can understand their own database.  Just like rails.  Extremely nice thing to have as a developer.

**Comment by light24bulbs** [comment]
Hey there, just wondering if there are any updates on this?  I'm starting to think I'd be interested in trying to do something myself. 

 It's pretty simple, all I really want is something that takes all the migrations and mashes them together by table so I can see what my tables look like.  Just exactly what rails does.

@rhys-vdw You asked what the purpose is, and for me it is just so I can see what my current data model is without reading through my migrations(like 40 on one project) and doing diffs in my head.  Running all the migrations in a batch is enough to set up a new database, for me anyway.  

You or @tgriesser mentioned that knex-migrate was getting a rewrite like a year ago or something and I was wondering if there is any movement on that.

**Comment by jcrben** [comment]
Related issue about "Auto generating migrations based on defined schema" https://github.com/tgriesser/knex/issues/815

**Comment by ramhejazi** [comment]
I was looking for a package with such features and unfortunately couldn't find one, so I created a simple migration library.
https://github.com/ramhejazi/dana
It's currently a beta software.

**Comment by ramhejazi** [comment]
I was looking for a package with, somehow, similar features and unfortunately couldn't find one, so I decided to create a simple migration library.
https://github.com/ramhejazi/dana
It's currently a beta software.

**Comment by MilesMorel** [comment]
Wondering if there's been any development around this? Seems like an interesting and potentially very useful idea

**Comment by kibertoad** [comment]
Nope. Feel free to take it up :)

**Comment by serv** [comment]
@kibertoad, I am interested in this feature and I want to clarify the requirements with you. What do you think about these requirements.

- Taking some pointers from [how Rails has done this](https://edgeguides.rubyonrails.org/active_record_migrations.html#types-of-schema-dumps), I see following data should be present in the schema dump.
  - batch: Display which the last batch updated the schema
  - tables
  - columns and their columnType
- We create a `schema.json` file. Here's an example.
```json
{
  "batch": 1,
  "tables": {
    "employees": {
      "name": "string",
      "title": "string",
      "salary": "integer" 
    },
    "teams": {
      "name": "string",
      "members": "INT[]"
    } 
  }
}, 
```
- `schema.json` is created and updated when migrations are ran and updates the database schema.
- `schema.json` is generated in `db` directory

**Comment by elhigu** [comment]
I think this is also one of those things, which should have their own repository and separate npm package where it is implemented.

**Comment by mercmobily** [comment]
Closing this in favour of #6234 which includes it

**Comment by factoidforrest** [comment]
9 years but you did it!! 🥳

**Comment by mercmobily** [comment]
Yeah... it took a while. We are going to focus on reorganising and triaging tickets first.  We will get there!
